Support increasing zero-ada outputs in balanceTx

Anviking opened the pull request:
- [x] Implement and test `WriteTx.computeMinimumCoinForTxOut`
- [x] Add `increaseZeroAdaOutputs`
    - Needed some ugly helpers in `Write.Tx`  module, as the ledger lenses are not in the version we currently use.
- [x] Some error messages in balanceTx goldens are affected by 44 lovelace (fee for 1 byte), I'm guessing as the pre-processing of the partial transaction before evaluating the fee normalizes it from Alonzo to Babbage. I can't see how this should be anything to worry about.


### Comments

<!-- Additional comments, links, or screenshots to attach, if any. -->

### Issue Number

ADP-2347

<!-- Reference the Jira/GitHub issue that this PR relates to, and which requirements it tackles.
  Note: Jira issues of the form ADP- will be auto-linked. -->

Anviking commented on lib/wallet/src/Cardano/Wallet.hs:
```diff
@@ -1633,6 +1639,43 @@ balanceTransaction ctx change pp ti wallet ptx = do
+    -> WriteTx.PParams (Cardano.ShelleyLedgerEra era)
+    -> Cardano.Tx era
+    -> Cardano.Tx era
+increaseZeroAdaOutputs era pp (Cardano.Tx cardanoBody keyWits) =
```
Would like to move along with `balanceTransaction` to some `Write.Tx.Balance` module*. I think it's better to merge this first, but I probably should get to it near-immediately afterwards.

*) Ultimately `increaseZeroAdaOutputs` should preferably be decoupled from `balanceTransaction`, but for the medium-term I think this should be fine.

Anviking commented on lib/wallet/src/Cardano/Wallet/Write/Tx.hs:
```diff
@@ -543,3 +704,11 @@ withStandardCryptoConstraint
+
+withCLIConstraint
+    :: RecentEra era
+    -> (CLI (ShelleyLedgerEra era) => a)
```
Could be added as a superclass to `RecentEra` after ADP-2353 ... or maybe now?

Anviking replied:
Going with later. Requires `IsRecentEra era =>` rather than `RecentEra era` for us to easily get the constraint without pattern-matching or calling a helper. As [ADP-2353](https://input-output.atlassian.net/browse/ADP-2353) may affect the choice between passing `IsRecentEra era =>` or value-level `RecentEra` to functions, it seems better to see how that plays out first.

Anviking commented on lib/wallet/src/Cardano/Wallet.hs:
```diff
@@ -629,6 +630,9 @@ import qualified Data.Set as Set
 import qualified Data.Vector as V
 
+import qualified Cardano.Ledger.Babbage.TxBody as Babbage
+import qualified Cardano.Ledger.Core as Ledger
```
Removing

Anviking commented on lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -2213,7 +2219,10 @@ mockProtocolParameters = dummyProtocolParameters
+    , minimumUTxO = minimumUTxOForShelleyBasedEra Cardano.ShelleyBasedEraBabbage
+        def
+            { Babbage._coinsPerUTxOByte = testParameter_coinsPerUTxOByte_Babbage
+            }
```
Needed to update this to exactly match the values `Cardano.ProtocolParameters`. Opted not to do anything nicer now, but we should switch to have `balanceTx` take a `Ledger.PParams era` only, rather than the current `(W.ProtocolParameters, Cardano.ProtocolParameters)`.

Anviking commented on lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -2455,19 +2464,22 @@ balanceTransactionSpec = describe "balanceTransaction" $ do
-        let rootK = Shelley.unsafeGenerateKeyFromSeed (mw, Nothing) mempty
+    it "increases zero-ada outputs to minimum" $ do
+        let era = WriteTx.RecentEraBabbage
+        let out = TxOut dummyAddr (TokenBundle.fromCoin (Coin 0))
```
I would like not to use wallet types here — and e.g. have a test case with a 10KB inline script, which should produce a larger minAda value than what was possible in Alonzo — but avoiding for now to keep scope of PR small.

Anviking commented on lib/wallet/src/Cardano/Wallet/Write/Tx.hs:
```diff
@@ -454,6 +524,65 @@ downcastTxOut (TxOutInRecentEra addr val Alonzo.NoDatum Nothing)
+    in
+        if actualCoin >= requiredMin
+        then Right ()
+        else Left $ requiredMin <-> actualCoin
```
🤔 First I didn't want to add heavy types (`expected: , actual:`) because it wasn't needed yet and returned a `Bool`. Then I needed some more details in the quickcheck property counterexample (currently the only caller of this function) and added the `shortfall`.

jonathanknowles replied:
Hmm.

Given that `evaluateMinLovelaceOutput` does not reach a fixed point, what should the caller do with the returned `Coin` deficit? It's not enough to simply compute `original <> deficit`, as this may cause an increase in the serialised length of the ada quantity, and then `evaluateMinLovelaceOutput` would presumably return an even higher value.

If the specified value is below the minimum, then why not return the result of `computeMinimumCoinForTxOut`? At least this way, the caller can choose any value greater than or equal to the returned value, and be confident that it'll work.

Returning a minimum that can always be safely increased is actually the behaviour of our existing API endpoints: I think it would be nice to preserve this behaviour!

To make the purpose of the returned `Coin` value more obvious, I would propose either putting it an a record with a descriptive field name, or embedding it in a newtype:
```hs
newtype SafeMinimum a = SafeMinimum {getSafeMinimum :: a}
```
Then the return type of the function could be:
```hs
Either (SafeMinimum Coin) ()
```

Anviking replied:
> Given that evaluateMinLovelaceOutput does not reach a fixed point

👍 Good point, I forgot about that here. 

Perhaps I'll leave the function returning `Bool` with a TODO to add a more descriptive error for the sake of 1) taking over responsibility from the checks in coin-selection 2) providing counterexample details in the property tests 🤔

Anviking replied:
69a9f4e43f7f0e00113f22f1f71c0a5158e1dde9

Anviking commented on lib/wallet/src/Cardano/Wallet.hs:
```diff
@@ -1905,6 +1919,16 @@ balanceTransactionWithSelectionStrategy
+            Just recentEra = WriteTx.toRecentEra $ Cardano.cardanoEra @era
+        in
+            PartialTx (increaseZeroAdaOutputs recentEra ledgerPParams tx) u rs
+
```
The amount of logic in these `where` clauses is starting to get a bit much, yes. I would like to move balanceTx to a separate module before attempting to split it up, though.

Anviking replied:
Although it would be enough to move the `increaseZeroAdaOutuputs` to `balanceTransaction` as opposed to `balanceTransactionWithSelectionStrategy`. The only problem is that the `balanceTransactionWithSelectionStrategy` name then makes less sense with both zero-ada and selection strategy differing, so perhaps `balanceTransactionInternal`? 🤨 not too much of a fan

Anviking replied:
Another option is to allow the caller of `W.balanceTransaction` to _also_ call `W.increaseZeroAdaOutputs` if/as they wish. `W.increaseZeroAdaOutputs` originates from the wallet HTTP API, and _might_ potentially be considered out-of-place or ad-hoc by Marlowe or other users.

The problem here is that we'd currently be unable to remove `guardZeroAdaOutputs`. We'd have to wait for full balanceTx re-use and removal of zero-ada-output-increasing in the underlying coin-selection.

Combining this latter point with the fact that we're not sure whether non-wallet users might dislike such a feature, I'd keep `W.increaseZeroAdaOutputs` in `W.balanceTransaction` for now.

jonathanknowles replied:
My vote would be for an outer orchestration function that composes `increaseZeroAdaOutputs` with `balanceTransaction`. Something like:

```hs
outerOrchestrationFn ptx = balanceTransaction (increaseZeroAdaOutputs ptx)
```

The outer function could also do any other validation / error checking of pre-conditions, if desired, so that `balanceTransaction` could already assume that its preconditions have been satisfied.

Then, we wouldn't need to add extra complexity to the test suite for `balanceTransaction` to cover the 0-ada output case, as `balanceTransaction` would never need to modify user-specified outputs.

Anviking replied:
Went with 

> Although it would be enough to move the increaseZeroAdaOutuputs to balanceTransaction as opposed to balanceTransactionWithSelectionStrategy. The only problem is that the balanceTransactionWithSelectionStrategy name then makes less sense with both zero-ada and selection strategy differing, so perhaps balanceTransactionInternal?

but with `balanceTransactionWithSelectionStrategyAndNoZeroAdaAdjustment` instead of `balanceTransactionInternal `.

jonathanknowles commented on lib/wallet/src/Cardano/Wallet.hs:
```diff
@@ -1657,12 +1672,11 @@ balanceTransactionWithSelectionStrategy
     (internalUtxoAvailable, wallet, _pendingTxs)
     selectionStrategy
-    ptx@(PartialTx partialTx inputUTxO redeemers)
+    ptx'
```
Suggestion: use something more descriptive than `'` to help readers distinguish between the unadjusted tx and the tx with adjusted outputs.

```suggestion
    ptxUnadjusted
```

Anviking replied:
https://github.com/input-output-hk/cardano-wallet/pull/3571/commits/f1cef4ceaa9d8a8bf03265ea8215d73d0e65cde9

jonathanknowles commented on lib/wallet/src/Cardano/Wallet.hs:
```diff
@@ -1905,6 +1919,16 @@ balanceTransactionWithSelectionStrategy
+                    nodePParams
+            Just recentEra = WriteTx.toRecentEra $ Cardano.cardanoEra @era
+        in
+            PartialTx (increaseZeroAdaOutputs recentEra ledgerPParams tx) u rs
```
I know you generally prefer `let` clauses over `where` clauses!

However in this case it seems to use quite a lot of horizontal and vertical space.

Using a `where` would save 2 lines and one indentation level:

```suggestion
    ptx@(PartialTx partialTx inputUTxO redeemers) =
        PartialTx (increaseZeroAdaOutputs recentEra ledgerPParams tx) u rs
      where
        PartialTx tx u rs = ptx'
        ledgerPParams = Cardano.toLedgerPParams
            (Cardano.shelleyBasedEra @era) nodePParams
        Just recentEra = WriteTx.toRecentEra $ Cardano.cardanoEra @era
```

Anviking replied:
Maybe here `where` makes sense. I'll see if I can remove the need for this at all though (move to different function)

Also: I should remove the `Just recentEra` pattern match...

Anviking replied:
https://github.com/input-output-hk/cardano-wallet/pull/3571/commits/f1cef4ceaa9d8a8bf03265ea8215d73d0e65cde9

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Write/Tx.hs:
```diff
@@ -37,19 +40,46 @@ module Cardano.Wallet.Write.Tx
+    , StandardCrypto
+    , StandardBabbage
+    , StandardAlonzo
+
```
```suggestion
```

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Write/Tx.hs:
```diff
@@ -254,6 +299,31 @@ unsafeMkTxIn hash ix = Ledger.mkTxInPartial
+modifyTxOutValue RecentEraBabbage f (Babbage.TxOut addr val dat script)=
+        withStandardCryptoConstraint RecentEraBabbage $ Babbage.TxOut addr (f val) dat script
+modifyTxOutValue RecentEraAlonzo f (Alonzo.TxOut addr val dat) =
+        withStandardCryptoConstraint RecentEraAlonzo $ Alonzo.TxOut addr (f val) dat
```
Request: keep lines within the limit / use 4-space indents, as per the standard.

```suggestion
  modifyTxOutValue RecentEraBabbage f (Babbage.TxOut addr val dat script) =
      withStandardCryptoConstraint RecentEraBabbage $
          Babbage.TxOut addr (f val) dat script
  modifyTxOutValue RecentEraAlonzo f (Alonzo.TxOut addr val dat) =
      withStandardCryptoConstraint RecentEraAlonzo $
          Alonzo.TxOut addr (f val) dat
```

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Write/Tx.hs:
```diff
@@ -454,6 +524,64 @@ downcastTxOut (TxOutInRecentEra addr val Alonzo.NoDatum Nothing)
+--
+-- NOTE:
+computeMinimumCoinForTxOut
+    :: forall era. RecentEra era -- FIXME [ADP-2353] Replace 'RecentEra' with 'IsRecentEra'
```
```suggestion
    :: forall era. RecentEra era
    -- ^ FIXME [ADP-2353] Replace 'RecentEra' with 'IsRecentEra'
```

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Write/Tx.hs:
```diff
@@ -454,6 +524,64 @@ downcastTxOut (TxOutInRecentEra addr val Alonzo.NoDatum Nothing)
+            modifyTxOutValue era $ modifyCoin (const $ toLedger txOutMaxCoin)
+
+isBelowMinimumCoinForTxOut
+    :: forall era. RecentEra era -- FIXME [ADP-2353] Replace 'RecentEra' with 'IsRecentEra'
```
```suggestion
    :: forall era. RecentEra era
    -- ^ FIXME [ADP-2353] Replace 'RecentEra' with 'IsRecentEra'
```

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Write/Tx.hs:
```diff
@@ -454,6 +524,64 @@ downcastTxOut (TxOutInRecentEra addr val Alonzo.NoDatum Nothing)
+        minCoin = computeMinimumCoinForTxOut era pp out
+    in
+        withStandardCryptoConstraint era $
+            modifyTxOutValue era (modifyCoin (const minCoin)) out
```
You can save 2 lines and 1 level of indentation here:

```suggestion
setMinimumCoinForTxOut era pp out = withStandardCryptoConstraint era $
    modifyTxOutValue era (modifyCoin (const minCoin)) out
  where
    minCoin = computeMinimumCoinForTxOut era pp out
```

Anviking replied:
I think this is a case where I do prefer sacrificing lines and indentation 😁. My thinking is that the definition of `minCoin` is important to understanding the rest of the implementation of `setMinimumCoinForTxOut`, so it should be read _before_ reading the next line(s).

Anviking replied:
Putting the `withStandardCryptoConstraint ` on the first line like you have might be an idea though...

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Write/Tx.hs:
```diff
@@ -491,9 +619,74 @@ utxoFromTxOutsInLatestEra
+--
+-- TODO [ADP-2353] Move to @cardano-api@ related module
+modifyLedgerBody
+    :: (Core.TxBody (ShelleyLedgerEra cardanoEra)
+        -> Core.TxBody (ShelleyLedgerEra cardanoEra))
+    -> Cardano.Tx cardanoEra
+    -> Cardano.Tx cardanoEra
```
Just an idea:
```suggestion
    :: txBody ~ Core.TxBody (ShelleyLedgerEra era)
    => (txBody -> txBody)
    -> (Cardano.Tx era -> Cardano.Tx era)
```

Anviking replied:
Will remove the type application in [ADP-2353](https://input-output.atlassian.net/browse/ADP-2353)

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Write/Tx.hs:
```diff
@@ -491,9 +619,74 @@ utxoFromTxOutsInLatestEra
+            auxData
+            validity
+    in
+        Cardano.Tx body' keyWits
```
The GADTs defined by the ledger are nice, but a real pain if we just want to update one field!

How nice it would be if we could write something like this:

```hs
modifyLedgerBody = over (#body . #ledgerBody)
```

Anviking replied:
I was just thinking yesterday about record dot syntax and how `HasField` constraints could be added as superclass to `RecentEra` for the ultimate ease-of-use.

But you're right, something similar could possibly be done today already...

Thanks for reminding me about `over` — I'll have a look!

Anviking replied:
I think we currently would need to define e.g.
```haskell
instance Data.Generics.Product.Fields.HasField' "value" (Babbage.TxOut StandardBabbage) (Mary.Value StandardCrypto) where
```
instances to make generic-lens work, which would be orphans. The ledger does contain a lot of `GHC.Records.HasField` instances, but those do not include setters.

**However**, on more recent `cardano-ledger` revisions it seems they have added `Generic` instances to lots of types. I think those generic instances together with turning `IsRecentEra era` into `class ManyHasFieldConstraints => IsRecentEra era where` would be what we'd need to remove the helper functions I added here.

Anviking replied:
I suspect we should not let the helpers block the merge of this PR, but it would be **very** nice to get this kind of thing working!

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Write/Tx.hs:
```diff
@@ -543,3 +736,12 @@ withStandardCryptoConstraint
     RecentEraBabbage -> a
     RecentEraAlonzo -> a
+
+withCLIConstraint
+    :: RecentEra era
+    -> (CLI (ShelleyLedgerEra era) => a)
```
No need to wrap this, AFAICT:

```suggestion
withCLIConstraint :: RecentEra era -> (CLI (ShelleyLedgerEra era) => a)
```

Anviking replied:
Keeping it for similarity with `withStandardCryptoConstraint`.

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Write/Tx.hs:
```diff
@@ -543,3 +736,12 @@ withStandardCryptoConstraint
+withCLIConstraint era a = case era of
+    RecentEraBabbage -> a
+    RecentEraAlonzo -> a
+
```
```suggestion
```

jonathanknowles commented on lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -2455,19 +2464,22 @@ balanceTransactionSpec = describe "balanceTransaction" $ do
-        let rootK = Shelley.unsafeGenerateKeyFromSeed (mw, Nothing) mempty
+    it "increases zero-ada outputs to minimum" $ do
+        let era = WriteTx.RecentEraBabbage
+        let out = TxOut dummyAddr (TokenBundle.fromCoin (Coin 0))
```
```suggestion
        let out = TxOut dummyAddr mempty
```

Anviking replied:
I think `(TokenBundle.fromCoin (Coin 0))` makes more sense over `mempty` here for the sake of visually highlighting the difference to `out'`:

```haskell
let out = TxOut dummyAddr (TokenBundle.fromCoin (Coin 0))
let out' = TxOut dummyAddr (TokenBundle.fromCoin (Coin 874_930))
```

jonathanknowles commented on lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -3795,11 +3812,42 @@ prop_balanceTransactionValid wallet (ShowBuildable partialTx) seed
+            let outs = F.toList $ Alonzo.outputs body
+                hasZeroAda (Alonzo.TxOut _ val _) = Value.coin val == Ledger.Coin 0
+            in
+                any hasZeroAda outs
```
```suggestion
      hasZeroAdaOutputs (Cardano.Tx (Cardano.ShelleyTxBody _ body _ _ _ _) _) =
          any hasZeroAda (Alonzo.outputs body)
        where
          hasZeroAda (Alonzo.TxOut _ val _) = Value.coin val == Ledger.Coin 0
```

Anviking replied:
👍

jonathanknowles commented on lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -3832,6 +3880,7 @@ prop_balanceTransactionValid wallet (ShowBuildable partialTx) seed
         TxOutValue _ val -> val
 
     (_, nodePParams) = mockProtocolParametersForBalancing
+    ledgerPParams = Cardano.toLedgerPParams Cardano.ShelleyBasedEraAlonzo nodePParams
```
```suggestion
    ledgerPParams = Cardano.toLedgerPParams
        Cardano.ShelleyBasedEraAlonzo nodePParams
```

jonathanknowles reviewed (commented):
Hi @Anviking 

Thanks for making this PR!

I've left a few initial comments.

Anviking commented on lib/wallet/src/Cardano/Wallet/Write/Tx.hs:
```diff
@@ -454,6 +525,64 @@ downcastTxOut (TxOutInRecentEra addr val Alonzo.NoDatum Nothing)
+-- prop> let
+-- prop>     c = computeMinimumCoinForUTxO out
+-- prop> in
+-- prop>     forall c' > c. satisfiesMinimumUTxO (out { coin = c' })
```
Need to tweak the formatting a bit.

And the property check coverage is actually lacking. I thought we'd get it for free from prop_balanceTransactionValid but this is not the case currently at least. Setting `maxCoin = 0` only makes the golden fail.

Regardless, `computeMinimumCoinForTxOut` should definitely be property tested on its own, and possibly `increaseZeroAdaOutputs` too.

Anviking replied:
Tweaked doc comment & added property for `computeMinimumCoinForTxOut`.

jonathanknowles commented on lib/wallet/src/Cardano/Wallet.hs:
```diff
@@ -1614,25 +1615,53 @@ balanceTransaction
+            else c
+
+-- | Internal helper to 'balanceTransaction'
+balanceTransactionWithSelectionStrategyAndNoZeroAdaAdjustment
```
Perhaps call this `balanceTransactionInternal`?

jonathanknowles commented on lib/wallet/src/Cardano/Wallet.hs:
```diff
@@ -1614,25 +1615,53 @@ balanceTransaction
+        flip (WriteTx.modifyTxOutCoin era) out $ \c ->
+            if c == mempty
+            then WriteTx.computeMinimumCoinForTxOut era pp out
+            else c
```
👍🏻

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Write/Tx.hs:
```diff
@@ -254,6 +299,33 @@ unsafeMkTxIn hash ix = Ledger.mkTxInPartial
+    -> TxOut (ShelleyLedgerEra era)
+    -> Value (ShelleyLedgerEra era)
+value RecentEraBabbage (Babbage.TxOut _ val _ _) = val
+value RecentEraAlonzo (Alonzo.TxOut _ val _) = val
```
This function only operates on values of type `TxOut`, but the name indicates a greater level of generality.

Perhaps rename it to reflect the specificity?

```suggestion
txOutValue
    :: RecentEra era
    -> TxOut (ShelleyLedgerEra era)
    -> Value (ShelleyLedgerEra era)
txOutValue RecentEraBabbage (Babbage.TxOut _ val _ _) = val
txOutValue RecentEraAlonzo (Alonzo.TxOut _ val _) = val
```

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Write/Tx.hs:
```diff
@@ -454,6 +526,60 @@ downcastTxOut (TxOutInRecentEra addr val Alonzo.NoDatum Nothing)
+        actualCoin = getCoin era out
+    in
+        actualCoin < requiredMin
+  where
```
Hmm. This is just my opinion (so take it with a pinch of salt), but I still reckon using `where` would be clearer here.

In the suggestion below, the most important part of the function definition, namely the test `actualCoin < requiredMin`, is right at the top: (If the reader wishes to delve into more details, they can carry on reading.)

It also uses less horizontal and vertical space (saving two lines).

```suggestion
    actualCoin < requiredMin
  where
    requiredMin = evaluateMinLovelaceOutput pp out
    actualCoin = getCoin era out
```

Anviking replied:
> the most important part of the function definition

🤔 Fair enough...

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Write/Tx.hs:
```diff
@@ -491,9 +617,74 @@ utxoFromTxOutsInLatestEra
+txBody RecentEraAlonzo = Alonzo.body
+
+-- Until we have convenient lenses to use
+outputs
+    :: RecentEra era
+    -> Core.TxBody (ShelleyLedgerEra era)
+    -> [TxOut (ShelleyLedgerEra era)]
+outputs RecentEraBabbage = map sizedValue . toList . Babbage.outputs
+outputs RecentEraAlonzo = toList . Alonzo.outputs
```
Since this function can only extract outputs from a `TxBody`:

```suggestion
txBodyOutputs
    :: RecentEra era
    -> Core.TxBody (ShelleyLedgerEra era)
    -> [TxOut (ShelleyLedgerEra era)]
txBodyOutputs RecentEraBabbage = map sizedValue . toList . Babbage.outputs
txBodyOutputs RecentEraAlonzo = toList . Alonzo.outputs
```

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Write/Tx.hs:
```diff
@@ -491,9 +617,74 @@ utxoFromTxOutsInLatestEra
 --------------------------------------------------------------------------------
 -- Compatibility
 --------------------------------------------------------------------------------
+
```
👍🏻

jonathanknowles commented on lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -2454,19 +2463,30 @@ balanceTransactionSpec = describe "balanceTransaction" $ do
+    it "increases zero-ada outputs to minimum" $ do
+        let era = WriteTx.RecentEraBabbage
+        let out = TxOut dummyAddr (TokenBundle.fromCoin (Coin 0))
+        let out' = TxOut dummyAddr (TokenBundle.fromCoin (Coin 874_930))
```
Hmm. Where did the `874_930` value come from?

Is it just picked randomly to be somewhat lower than the minimum required value?

Anviking replied:
This is the computed (safe) minimum. Using the ~alonzo computation.~

Anviking replied:
No, right this is Babbage. Because it's lower than what is seen in [`MinimumUTxOSpec`](https://github.com/input-output-hk/cardano-wallet/blob/d59b05e6e7b2c0b1d114f1bc87b737fd1385d8c7/lib/wallet/test/unit/Cardano/Wallet/Shelley/MinimumUTxOSpec.hs#L493) because the dummy address here only contains a payment key (without the extra 28 bytes for delegation key).

```
995610-(28*34482/8) = 874923
```

jonathanknowles commented on lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -2454,19 +2463,30 @@ balanceTransactionSpec = describe "balanceTransaction" $ do
 
+        head outs `shouldBe` (toBabbageTxOut out' Nothing)
+
+
```
```suggestion

```

jonathanknowles commented on lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -3662,6 +3684,8 @@ prop_balanceTransactionValid wallet (ShowBuildable partialTx) seed
                     $ classify (txFee tx > Cardano.Lovelace 1_000_000)
                         "fee above 1 ada"
+                    $ classify (hasZeroAdaOutputs $ view #tx partialTx)
+                        "partial tx had zero ada outputs"
```
I'm guessing that since we're using `classify`, that we currently don't actually **_require_** that this case actually hits during a given test run?

Is there some way to use `checkCoverage` and `cover` to ensure that we hit this case?

Anviking replied:
I know. Maybe I'll get to it after fixing the `PParam era` vs `W.ProtocolParameters era` and polishing the tests a bit.

For the immediate time, coverage in this property shouldn't matter too much when we have the `computeMinimumCoinForTxOut` properties, a `balanceTx` unit golden, and — soon — integration test coverage through constructTx.

jonathanknowles commented on lib/wallet/test/unit/Cardano/Wallet/Write/TxSpec.hs:
```diff
@@ -143,6 +152,28 @@ spec = do
 
+    describe "TxOut" $ do
+        describe "computeMinAdaQuantity" $ do
+            it "isBelowMinimumCoinForTxOut (setCoin (result <> delta)) \
+               \ == False (Alonzo)"
```
I can't help wondering if these long test name strings wouldn't be better as a comment on a separate, named property. Writing them as strings with line continuation characters makes it quite awkward to edit and format, IMO.

Would it be easy to extract these out into separate, named properties?

Anviking replied:
I'm fond/curious of keeping the property implementation close to the `it` description, and not having to maintain the function name and optional doc comment. So I would prefer to continue seeing how this turns out.

jonathanknowles commented on specifications/api/swagger.yaml:
```diff
@@ -2759,7 +2759,7 @@ components:
         burn: *ApiAssetMintBurn
         validity_interval: *ApiValidityIntervalExplicit
-        script_integrity: *scriptIntegrityHash 
+        script_integrity: *scriptIntegrityHash
```
I wonder who is leaving these stray trailing spaces everywhere. I also deleted the exact same space in the exact same file in one of my PRs. 😄

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -288,7 +288,6 @@ import qualified Data.Sequence.Strict as StrictSeq
 import qualified Data.Set as Set
 import qualified Data.Text as T
 
-
```
👍🏻

jonathanknowles commented on specifications/api/swagger.yaml:
```diff
@@ -4727,20 +4727,6 @@ x-errBalanceTxUnderestimatedFee: &errBalanceTxUnderestimatedFee
-        to increase the values to the minimum allowed ada value.
-    code:
-      type: string
-      enum: ["balance_tx_zero_ada_output"]
```
👍🏻

jonathanknowles reviewed (approved):
Hi @Anviking 

Thanks again for making this PR. It looks good to me. I have a few comments and suggestions.

My only real reservation is the long function name `balanceTransactionWithSelectionStrategyAndNoZeroAdaAdjustment`, which attempts to capture the nature of an internal function composition in the name. Perhaps we could rename this to something like `balanceTransactionInternal`?

Anviking committed:
Fix incompatible mock PParam values

Fixes property test failure `--match "balanceTransaction" --seed 18778639` due
to one part of the code using the Alonzo computation and one part the Babbage
one. `balanceTx` taking both `W.ProtocolParamters` and
`Cardano.ProtocolParamters` is unfortunately, and we/I should get to replacing
them with `Ledger.PParams era` soon.

Unfortunately we also need to disable the balanceTx unit test for 0-ada outputs
in the meantime.

Anviking committed:
Implement `WriteTx.computeMinimumCoinForTxOut`

Anviking committed:
Add/support `increaseZeroAdaOuts` in `balanceTx`

This allows us to remove the `ErrBalanceTxZeroAdaOutput` "not yet
implemented" error.

Anviking committed:
fixup: regenerate ApiError.json
